### ISSUE TITLE
Entrypoint Image Lookup does not use PodTemplate ImagePullSecrets

### ISSUE THREAD

**Issue by skaegi** [tektoncd/pipeline#3799]
At this point this is just by looking at the code and not validated by me. (although reported by a user)
Current main -- https://github.com/tektoncd/pipeline/blob/e2f528b0b247efb3b52b10800aad4ffd9046a266/pkg/pod/entrypoint_lookup_impl.go#L64

We're creating the k8schain with the service account, but also should add in the image pull secrets referenced in the pod template.

**Comment by wlynch** [comment]
/remove-lifecycle rotten I think this is the root cause of https://github.com/tektoncd/pipeline/issues/2316 - probably worth doing to make sure we're using consistent credentials.

**Comment by wlynch** [comment]
@sbwsg I know you've been looking at related k8schain code recently. cc-ing you in case you have any interest in this issue.

**Comment by concaf** [comment]
I believe that the certificates mounted via `SSL_CERT_DIR` are not respected either.

**Comment by Yongxuanzhang** [comment]
I'm a member of tektoncd but why I cannot be assigned?

**Comment by vdemeester** [comment]
@Yongxuanzhang you have to comment once, or be the one commenting `/assign` to get assigned 🙃

**Comment by Yongxuanzhang** [comment]
> @Yongxuanzhang you have to comment once, or be the one commenting `/assign` to get assigned 🙃

okok, thanks!

**Comment by imjasonh** [comment]
We'll need to make sure that users give the Tekton controller SA permissions to use those secrets too.

**Comment by Yongxuanzhang** [comment]
> 

Hi @imjasonh, thank you so much for the comment! I'm quite new to this, could you elaborate more on this? Do you mean we should document it to tell users to give controller SA permissions? 
And to give the controller the permissions, we need to use Rolebinding to grant the controller access to secrets, or can we add imagePullSecrets to the controller's SA? 
Thanks!

**Comment by Yongxuanzhang** [comment]
> 

Hi @imjasonh, 
Thank you so much for the detailed explanations!! ❤️ 
I have 1 question regarding this, 

I wonder how it works before, if we don't pass the `imagePullSecrets` explicitly to `k8schain.New`, it will also [search](https://github.com/google/go-containerregistry/blob/d9bfbcb99e526b2a9417160e209b816e1b1fb6bd/pkg/authn/k8schain/k8schain.go#L80) the  `imagePullSecrets` attached to the SA. 
And in our [development guide](https://github.com/tektoncd/pipeline/blob/766fb259b517534653bc45f41e5ae6dc8f9a60b7/DEVELOPMENT.md?plain=1#L234), we need to configure `imagePullSecrets` for `controller` and `webhook`'s SA.  

So are the following 2 things equivalent if users just want to pull images? 
1. Introduce this change, set the `imagePullSecrets` in pod spec and use rolebinding to grant SA the access to `secrets` like you mentioned in last comment;
2. Like the [development guide](https://github.com/tektoncd/pipeline/blob/766fb259b517534653bc45f41e5ae6dc8f9a60b7/DEVELOPMENT.md?plain=1#L234), directly set the `imagePullSecrets` for `controller` and `webhook` 's SA. 

If so do we really need to introduce this change? What will be the benefits of this? 
If not, what is the major difference of these 2 ways?
Thank you again for your help!

**Comment by imjasonh** [comment]
> So are the following 2 things equivalent if users just want to pull images?

Mostly. However, a SA's attached `imagePullSecrets` must exist in that SA's namespace. So if the `podTemplate`'s `imagePullSecrets` exist in `my-ns`, the Tekton controller SA's `imagePullSecrets` (in the `tekton-pipelines` namespace) won't be able to reference it as a (SA-defined) `imagePullSecret`.

> If so do we really need to introduce this change? What will be the benefits of this?

This change would let users specify and grant access to other secrets that the SA should use to authorize image pulls for entrypoint lookup. Those secrets could then also exist in a non-Tekton namespace.

I'll leave it to @skaegi to say how important this change is. From the initial comment it sounds like it may just be a nice-to-have, and there hasn't exactly been an avalanche of comments describing a need for this change -- maybe https://github.com/tektoncd/pipeline/issues/2316 is something worth looking into more.

**Comment by Yongxuanzhang** [comment]
> > So are the following 2 things equivalent if users just want to pull images?
> 
> Mostly. However, a SA's attached `imagePullSecrets` must exist in that SA's namespace. So if the `podTemplate`'s `imagePullSecrets` exist in `my-ns`, the Tekton controller SA's `imagePullSecrets` (in the `tekton-pipelines` namespace) won't be able to reference it as a (SA-defined) `imagePullSecret`.
> 
> > If so do we really need to introduce this change? What will be the benefits of this?
> 
> This change would let users specify and grant access to other secrets that the SA should use to authorize image pulls for entrypoint lookup. Those secrets could then also exist in a non-Tekton namespace.
> 
> I'll leave it to @skaegi to say how important this change is. From the initial comment it sounds like it may just be a nice-to-have, and there hasn't exactly been an avalanche of comments describing a need for this change -- maybe #2316 is something worth looking into more.

Got it! Thank you very much!

**Comment by Yongxuanzhang** [comment]
Hi @imjasonh , sorry to bother you again 😄 . I took some time to setup related env and test, read the code and I'm not sure the issue you mentioned will occur. Just correct me if I get it wrong: 
The params of `k8schain.New` are passed from https://github.com/tektoncd/pipeline/blob/0040bcaa9ab911b8b141699e94fee303e2134f63/pkg/pod/pod.go#L164
So it will use the `taskrun`'s namespace and SA instead of the Tekton controller's?

I also deploy a `taskrun` specified as below to my draft pr's branch, created a secret in `my-ns`, and there are no errors calling `k8schain.New` to fetch the secrets.

task_run.yaml:

```yaml
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  name: echo-hello-world-task-run
  namespace: my-ns
spec:
  taskRef:
    name: echo-hello-world
  podTemplate:
    imagePullSecrets:
      - name: secret-tekton
```
task.yaml:

```yaml
apiVersion: tekton.dev/v1beta1
kind: Task
metadata:
  name: echo-hello-world
  namespace: my-ns
spec:
  steps:
    - name: echo
      image: localhost:5000/ubuntu
```

**Comment by imjasonh** [comment]
Right you are! I forgot that the Tekton controller SA has read-only access to all secrets in the cluster, which makes this lookup succeed:

https://github.com/tektoncd/pipeline/blob/0040bcaa9ab911b8b141699e94fee303e2134f63/config/200-clusterrole.yaml#L58-L61

Sounds like this should be pretty straightforward to implement, the only remaining question I have is whether users may be depending on the current behavior where `imagePullSecrets` on the PodTemplate are ignored when looking up the entrypoint. It's unlikely, but someone might not expect this. Should we enable this behind a feature flag so users can opt-in for a release or two before making it the default behavior?

**Comment by Yongxuanzhang** [comment]
> Right you are! I forgot that the Tekton controller SA has read-only access to all secrets in the cluster, which makes this lookup succeed:
> 
> https://github.com/tektoncd/pipeline/blob/0040bcaa9ab911b8b141699e94fee303e2134f63/config/200-clusterrole.yaml#L58-L61
> 
> Sounds like this should be pretty straightforward to implement, the only remaining question I have is whether users may be depending on the current behavior where `imagePullSecrets` on the PodTemplate are ignored when looking up the entrypoint. It's unlikely, but someone might not expect this. Should we enable this behind a feature flag so users can opt-in for a release or two before making it the default behavior?

Thank you! My previous understanding is wrong but now is corrected. 
Since controller's SA may or may not have this access, I will document the rolebinding part in the PR.

**Comment by Yongxuanzhang** [comment]
Probably  forgot to link [PR](https://github.com/tektoncd/pipeline/pull/4496) to this issue

**Comment by pritidesai** [comment]
This should be fixed by #4496 which will be part of Pipelines 0.33 🎉 
@skaegi please feel free to reopen if you run into any issues. Thanks! 🙏
